application/models/Question.php

```diff
@@ -508,7 +508,7 @@ public function deleteAllSubquestions($exceptIds = [])
             ? array_diff($this->allSubQuestionIds, $exceptIds)
             : $this->allSubQuestionIds;
         $qidsCriteria = (new CDbCriteria())->addInCondition('qid', $ids);
-        $res = Question::model()->deleteAll((new CDbCriteria())->addInCondition('qid', $ids));
+        $questions = Question::model()->deleteAll((new CDbCriteria())->addInCondition('qid', $ids));
```

Comment:
due to referential integrity relationships have to be deleted BEFORE the model, so afterDelete() is not the correct place for deleting relations like L10n, Attribute,QuotaMember

we also cannot deleteByPk() because `afterModelDelete` will not be triggered

Reply:
> due to referential integrity relationships have to be deleted BEFORE the model, so afterDelete() is not the correct place for deleting relations like L10n, Attribute,QuotaMember

Are you sure with this ? Else we can replace by own delete : https://github.com/LimeSurvey/LimeSurvey/blob/593c68659798007d231687ecba67e8e2dd8becb0/application/models/Survey.php#L248


> we also cannot deleteByPk() because `afterModelDelete` will not be triggered

No : deleteByPk trigger afterModelDelete , i already tested some year before…

deleteByPk trigger onAfterDelete by Yii

Reply:
Oups, i dodn't see we are already in a deletion …

the 2 solution :  

1. remove delete function and move to afterDelete (i make some test if needed)
2. in this function : 
      - delete only the question 
      - if ok related QuestionAttribute etc …
      - for each subquestion : send `delete('qid :=subqid')`

Reply:
> 2. in this function :
>    
>    * delete only the question
>    * if ok related QuestionAttribute etc …
>    * for each subquestion : send `delete('qid :=subqid')`

I think this approach is the best.
Even more, enclose in a DB Transaction

Reply:
> Even more, enclose in a DB Transaction

I don't understand ?

Reply:
> I mean would be good that all DB delete operations get into a DB Transaction.

Great idea, but start without maybe.